src/lib.rs

```diff
@@ -327,53 +338,36 @@ where
     pub fn peek_iter(&self) -> PeekIter<Key, Value> {
         PeekIter {
             map_iter: self.map.iter(),
-            lru_cache: self,
+            lru_cache_ttl: self.time_to_live,
         }
     }
 
-    fn has_expiry(&self) -> bool {
-        self.time_to_live != Duration::new(std::u64::MAX, 999_999_999)
-    }
-
-    fn expired<Q: ?Sized>(&self, key: &Q) -> bool
-    where
-        Key: Borrow<Q>,
-        Q: Ord,
-    {
-        let now = Instant::now();
-        self.has_expiry()
-            && self
-                .map
-                .get(key)
-                .map_or(false, |v| v.1 + self.time_to_live < now)
-    }
-
-    fn remove_oldest_element(&mut self) {
-        let _ = self
-            .list
-            .pop_front()
-            .map(|key| assert!(self.map.remove(&key).is_some()));
-    }
-
-    fn check_time_expired(&self) -> bool {
-        self.has_expiry() && self.list.front().map_or(false, |key| self.expired(key))
-    }
-
     // Move `key` in the ordered list to the last
-    fn update_key<Q: ?Sized>(list: &mut VecDeque<Key>, key: &Q)
+    fn update_key<Q: ?Sized>(list: &mut Vec<Key>, key: &Q)
     where
         Key: Borrow<Q>,
         Q: Ord,
     {
         if let Some(pos) = list.iter().position(|k| k.borrow() == key) {
-            let k = list.remove(pos).unwrap();
-            list.push_back(k);
+            let (_, tail) = list.split_at_mut(pos);
+            tail.rotate_left(1);
         }
     }
 
     fn remove_expired(&mut self) {
-        while self.check_time_expired() {
-            self.remove_oldest_element();
+        let (map, list) = (&mut self.map, &mut self.list);
+        if let Some((i, val)) = self.time_to_live.and_then(|ttl| {
+            list.iter()
+                .enumerate()
+                .filter_map(|(i, key)| map.remove(key).map(|val| (i, val)))
+                .filter(|&(_, (_, t))| t + ttl >= Instant::now())
+                .next()
```

Comment:
Could this just be `find` instead of `filter(…).next()`?

Reply:
Yep